(PUP-9049)(PUP-3021) Refactor AIX user/group providers + fix KeyValue property

ekinanp opened the pull request:
The AIX user/group provider code was messy, hard to reason about, buggy, and not very idiomatic. This PR refactors them to be easier to understand. The Key/Value property also had some bugs that needed to be fixed.

This PR resolves the following tickets:

* PUP-9049
* PUP-3021
* PUP-7241
* PUP-7533
* PUP-7604

ekinanp commented:
jenkins please test this on aix61-POWERfa,aix71-POWERfa,aix72-POWERfa

branan commented:
Looks like the test failures are a Ruby 2.0/1.9.3 compatibility issue (`to_h` on arrays?) plus a few warnings and rubocop errors. Should be easy to clean up

puppetcla commented:
CLA signed by all contributors.

branan commented on lib/puppet/property/keyvalue.rb:
```diff
@@ -53,11 +72,33 @@ def process_current_hash(current)
+        @changed_or_new_keys ||= should_value.keys.select do |key|
+          # A key that's exclusively in should was added, so it is
+          # automatically changed.
+          next true unless current.key?(key)
```
I'm not familiar with this `next true...` syntax. What does it do?

branan replied:
Oh, I think I get it. It's like an early return for a block? That's gross

ekinanp replied:
Yeah, next is not a very good name.

smcclellan replied:
`!current.key?(key) || current[key] != should_value[key]` maybe?

ekinanp replied:
Yes, that is better. I realized my NOTE is contradicting itself (the first one) since the first part is saying there's a difference between current and should. Will reword that.

branan commented on lib/puppet/type/group.rb:
```diff
@@ -154,17 +154,15 @@ def is_to_s(currentvalue)
       desc "Specify group AIX attributes, as an array of `'key=value'` strings. This
         parameter's behavior can be configured with `attribute_membership`."
 
+      self.log_only_changed_or_new_keys = true
```
Given most other type/provider/parameter things are done in a more meta-programmed style, is it possible to define this in a way that this line could be `log_only_changed_or_new_keys true`

ekinanp replied:
Oh yeah, that was a mistake on my part. Thanks for pointing it out.

ekinanp replied:
Unfortunately, looks like I can't do that without failing rubocop (on a `warning, unused variable assignment` message). So will need to include the self here.

ekinanp replied:
Ah, that did not work b/c `attr_accessor` creates a `log_only_changed_or_new_keys` getter method.

ekinanp commented:
I'm still working on some small improvements to the PR while I'm fixing the CI failures. The overall structure won't change though.

smcclellan commented on lib/puppet/provider/aix_object.rb:
```diff
@@ -0,0 +1,390 @@
+      # Define the Getter methods for each of our properties + the attributes
+      # property
+      properties = [:attributes]
+      properties += mappings[:aix_attribute].keys
```
Seems like code smell to me to have all this generic code (dynamically created setters/getters, for example) mixed into the same file as AIX-specific code. Is this code at all useful at a more generic level, for other platforms?

ekinanp replied:
It's a common idiom in Puppet types/providers to do this. See for example:

https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/provider/user/directoryservice.rb#L25
https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/provider/parsedfile.rb#L162
https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider.rb#L414

The reason is b/c a property's `retrieve` and assignment methods delegate to these getters and setters first before anything else -- see https://github.com/puppetlabs/puppet/blob/master/lib/puppet/property.rb#L470 and https://github.com/puppetlabs/puppet/blob/master/lib/puppet/property.rb#L506 and https://github.com/puppetlabs/puppet/blob/master/lib/puppet/property.rb#L188

Personally, I like this kind of meta-programming. It's a nice, clean way of capturing the high level behavior that's going on, especially in this class where we see that all properties mapping to AIX attributes will be retrieved and set the same way. For this case, I do not think this is a code-smell.

smcclellan replied:
How much of the code in this file is only useful for AIX? I'd like this file to be as small as possible, so we can see what is truly different from platform to platform in files like these. (Essentially, make files matching "*aix*" as straightforward as possible, and move any meta-programming code into a different file that other platforms can use.)

smcclellan replied:
In other words, could any of this code move up into `Puppet::Provider`?

ekinanp replied:
Not in an obvious way, no. Puppet::Provider already has a default `mk_resource_methods` implemention here https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider.rb#L414 which is different from how we're using it.

ekinanp commented:
jenkins please test this on aix61-POWERfa,aix71-POWERfa,aix72-POWERfa

ekinanp commented:
jenkins please test this on aix61-POWERfa,aix71-POWERfa,aix72-POWERfa

ekinanp commented:
PR tests passed. This should be ready for review now.

ekinanp commented:
Updated the PR to group `mappings` and `numeric_mappings` separately in the User and Group providers. Also added a new class-level method to AixObject `function` that takes a `pure` function and converts it to an `impure` one (i.e. one that takes the provider instance as a param.). The comments indicate why this was done.

I'm going to write a doc. explaining the refactor to prep for the PR review meeting tomorrow.

ekinanp commented:
jenkins please test this on aix61-POWERfa,aix71-POWERfa,aix72-POWERfa

ekinanp commented:
jenkins please test this on aix61-POWERfa,aix71-POWERfa,aix72-POWERfa

ekinanp commented:
jenkins please test this on aix61-POWERfa,aix71-POWERfa,aix72-POWERfa

speedofdark commented on acceptance/tests/resource/group/should_manage_attributes_aix.rb:
```diff
@@ -0,0 +1,24 @@
+  confine :to, :platform => /aix/
+  
+  tag 'audit:medium',
+      'audit:refactor',  # Use block style `test_run`
```
Why didn't you change these to block style per the refactor comment?

ekinanp replied:
I think it is in block style -- just forgot to delete that comment when I copy-pasted an old test file. But definitely worth bringing it up during PR review. Thanks for catching this!

ekinanp replied:
Need to remove `audit:refactor`.

ekinanp commented on acceptance/lib/puppet/acceptance/aix_util.rb:
```diff
@@ -0,0 +1,169 @@
+        attributes.map { |attribute, value| "#{attribute}=#{value}" }
+      end
+
+      def assert_object_attributes_on(agent, object_get, object, expected_attributes)
```
TODO: Change `agent` => `host`

ekinanp commented on acceptance/lib/puppet/acceptance/aix_util.rb:
```diff
@@ -0,0 +1,169 @@
+MANIFEST
+      end
+
+      def run_attribute_management_tests(object_resource, id_property, attributes, changed_attributes)
```
TODO: Change `attributes` => `default_attributes`

ekinanp commented on acceptance/lib/puppet/acceptance/aix_util.rb:
```diff
@@ -0,0 +1,169 @@
+            )
+     
+            apply_manifest_on(agent, manifest) do |result|
+              assert_match(/Error:.*attributes.*setting.*'#{id_property}' properties.*'id' attributes/, result.stderr, "specifying a Puppet property as part of an AIX attribute should have errored, but received #{result.stderr}")
```
Simplify the regex to something like `Error:.*id_property.*id_attribute`

branan replied:
Maybe just `/Error:.*#{id_property}/`

ekinanp commented on acceptance/lib/puppet/acceptance/aix_util.rb:
```diff
@@ -0,0 +1,169 @@
+            assert_object_attributes_on(agent, object_get, name, attributes)
+          end
+
+          step "Ensure that Puppet accepts a hash for the attributes property" do
```
TODO: Put PUP-3021 here.

ekinanp commented on acceptance/lib/puppet/acceptance/aix_util.rb:
```diff
@@ -0,0 +1,169 @@
+          step "Ensure that Puppet accepts a hash for the attributes property" do
+            # We want to see if Puppet will actually do something with the
+            # attributes property, so let's set one of our attributes back
+            # to its old value.
```
Add a comment about why we need to test that we can pass in a hash for the attributes property.

ekinanp commented on acceptance/tests/resource/user/should_manage_attributes_aix.rb:
```diff
@@ -0,0 +1,27 @@
+  confine :to, :platform => /aix/
+  
+  tag 'audit:medium',
+      'audit:refactor',  # Use block style `test_run`
```
TODO: Remove `audit:refactor`

ekinanp commented on acceptance/tests/resource/user/should_manage_attributes_aix.rb:
```diff
@@ -0,0 +1,27 @@
+  attributes = {
+    'nofiles'       => 10000,
+    'fsize'         => 100000,
+    'data'          => '60000',
```
TODO: Fix this to be a number (60000, not '60000').

ekinanp commented on lib/puppet/property/keyvalue.rb:
```diff
@@ -12,9 +12,20 @@ class Property
     # @todo The node with an important message is not very clear.
     #
     class KeyValue < Property
```
TODO: Double check that the KeyValue property preserves redacting of Sensitive values. (NOTE: Independent of this PR, this is just to make sure we don't have an underlying bug here).

ekinanp replied:
@branan Yup, I've verified using both the `keyvalue` array and the `hash` value that Sensitive values are correctly redacted. Here's a sample output:
```
[0] [AIX] root@aix71-5:~ # /opt/puppetlabs/puppet/bin/puppet apply MANIFESTS/pup_3021.pp
Notice: Compiled catalog for aix71-5.delivery.puppetlabs.net in environment production in 0.03 seconds
Notice: /Stage[main]/Main/User[USER_15]/attributes: changed [redacted] to [redacted]
Notice: Applied catalog in 0.35 seconds
```

ekinanp commented on lib/puppet/property/keyvalue.rb:
```diff
@@ -53,11 +72,24 @@ def process_current_hash(current)
+        # changed_or_new_keys once (since this property will only be synced once).
+        #
+        @changed_or_new_keys ||= should_value.keys.select do |key|
+          ! current.key?(key) or current[key] != should_value[key]
```
TODO: Use `||` instead of `or`.

ekinanp commented on lib/puppet/property/keyvalue.rb:
```diff
@@ -84,12 +116,42 @@ def retrieve
+      # Note that for an array property value, the 'value' passed into the
+      # block corresponds to the array element.
+      validate do |value|
+        unless value.is_a?(String) or value.is_a?(Hash)
```
TODO: Change `or` to `||`

ekinanp commented:
Updated the PR addressing feedback from the PR review. The acceptance test stuff I forgot to put in a separate commit (ended up ammending it to the acceptance test commit instead. Sorry about that). The acceptance test updates are:

(1) Change `agent` => `host` in `assert_object_attributes_on`.
(2) Change `attributes` => `initial_attributes` in the attribute management tests. I introduced a variable `current_attributes` to make that represent the current attributes we're managing. Should make things more readable.
(3) Changed the error regex to `Error:.*'#{id_property}'.*'id'` (I'd still like to ensure that our error message prints out the attribute as well as the Puppet property).
(4) Clarified the step where we're testing to ensure that the attributes can be passed as a `hash` a bit more (this was the one where we discussed whether we should include ticket numbers or not in the codebase).
(5) Removed `audit:refactor`.
(6) Fixed the '60000' value used in the `data` attribute in the AIX user attribute management tests to be 60000, the number instead.

I also addressed using `||` instead of `or` in the KeyValue property. That one I did remember to put in a separate commit to squash later, so you should be able to view it.

ekinanp commented:
I updated the PR to use instance-level mappings. The code's in a separate commit that will be squashed once this PR's approved and ready for merge.

ekinanp commented:
jenkins please test this on aix61-POWERfa,aix71-POWERfa,aix72-POWERfa

ekinanp commented on lib/puppet/provider/aix_object.rb:
```diff
@@ -0,0 +1,461 @@
+    def split_list(list, sep)
+      return [""] if list.empty?
+
+      # If list is our sep., then that means it held two
```
I did know about that. Thanks!

ekinanp commented:
Updated the PR to incorporate @McdonaldSeanp 's feedback.

ekinanp commented:
jenkins please test this on aix61-POWERfa,aix71-POWERfa,aix72-POWERfa

ekinanp commented:
Added the Blocked label to remind myself to squash the "SQUASH" commits before the final merge. Otherwise, this PR's good to go.
